File: python/src/azoth/eos/reference/ge_uniquac_phase.py
"""``eos.ge_uniquac_phase`` - the fugacity coefficients of a UNIQUAC liquid.

Spec: ``specs/models/eos/ge_uniquac_phase.toml``. A *direct* model: no iteration, so no
algorithm block.

This is the pure-Python reference: a second, independent expression of the same
physics as the Rust kernel, held to it by ``test_cross_impl.py``.
"""

from __future__ import annotations

import math
from collections.abc import Sequence

from azoth import _models_gen
from azoth.core.errors import InvalidInputError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import GeUniquacPhaseResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.components import GeUniquacPhaseParameters
from azoth.eos.reference._ge_phase import ge_fugacities
from azoth.eos.reference.uniquac_activity_coefficients import uniquac_activity_coefficients

MODEL_ID = "eos.ge_uniquac_phase"


def ge_uniquac_phase(
    params: GeUniquacPhaseParameters,
    T: Q,
    P: Q,
    x: Sequence[float],
    aij: Sequence[Sequence[Q]],
) -> GeUniquacPhaseResult:
    """The fugacity coefficients of a liquid whose non-ideality is UNIQUAC's.

    ``phi_i = gamma_i P0_i / P``, which is NeqSim's ``ComponentGE.fugcoef`` for this
    phase. ``gamma_i`` comes from UNIQUAC on ``params``' ``r`` and ``q`` with ``aij``;
    ``P0_i`` is Antoine's correlation on the component's own coefficients at ``T``.

    ``aij`` is the caller's - no upstream table carries a UNIQUAC interaction matrix. The
    public wrapper resolves it from a keycard when it is omitted; this reference is handed
    it.

    Args:
        params: the resolved phase parameters, by component.
        T: absolute temperature.
        P: absolute pressure.
        x: the liquid's mole fractions; non-negative and summing to one.
        aij: the interaction matrix, in kelvin, directional with a zero diagonal.

    Returns:
        The activity coefficients, their logarithms, the fugacity coefficients and the
        pure-component saturation pressures.

    Raises:
        InvalidInputError: if ``params`` and ``x`` disagree in length, if ``x`` is not a
            composition, or if ``aij`` is not an ``N x N`` matrix.
        OutOfRangeError: if ``T`` or ``P`` is not positive.

    Example:
        >>> import azoth
        >>> from azoth.eos.components import ge_uniquac_phase_parameters
        >>> q = azoth.ureg.Quantity
        >>> r = ge_uniquac_phase(
        ...     ge_uniquac_phase_parameters(["methanol", "water"]), q(298.15, "K"),
        ...     q(100000.0, "Pa"), [0.5, 0.5],
        ...     aij=[[q(0.0, "K"), q(-71.0, "K")], [q(209.0, "K"), q(0.0, "K")]],
        ... )
        >>> round(r.gamma[0], 12)
        1.218544184873
    """
    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    t_si = input_to_si(spec, "T", T)
    p_si = input_to_si(spec, "P", P)
    apply_checks(checks.on_input, {"T": t_si, "P": p_si}.get, warnings)

    x = list(x)
    n = len(x)
    if n == 0:
        raise InvalidInputError("x", "a mixture of zero components has no fugacity coefficient")
    if len(params.antoine_type) != n:
        raise InvalidInputError(
            "components",
            f"the phase has {len(params.antoine_type)} component(s) but `x` has {n} entries",
        )
    if any(value < 0.0 for value in x):
        bad = next(i for i, value in enumerate(x) if value < 0.0)
        raise InvalidInputError("x", f"x[{bad}] is {x[bad]} but a mole fraction cannot be negative")
    if abs(sum(x) - 1.0) > 1.0e-09:
        raise InvalidInputError(
            "x",
            f"the mole fractions sum to {sum(x)}, not to one. Renormalising them here "
            "would make a composition error invisible in every number downstream, so "
            "it is refused instead",
        )

    if len(aij) != n or any(len(row) != n for row in aij):
        raise InvalidInputError(
            "aij",
            f"the interaction matrix is {len(aij)} x "
            f"{len(aij[0]) if aij else 0} but the mixture has {n} components",
        )
    # The activity coefficients, from the same r and q and the same interaction matrix
    # `eos.uniquac_activity_coefficients` uses - resolved through the same function, so the
    # two cannot disagree.
    from azoth.eos.components import UniquacParameters

    activity = uniquac_activity_coefficients(
        UniquacParameters(r=params.r, q=params.q), T, x, aij=aij
    )
    warnings.extend(activity.warnings)
    gamma = list(activity.gamma)

    saturated = ge_fugacities(gamma, params, t_si, p_si)
    warnings.extend(saturated.warnings)

    ln_gamma = [math.log(value) for value in gamma]

    return GeUniquacPhaseResult(
        gamma=tuple(gamma),
        ln_gamma=tuple(ln_gamma),
        ln_phi=tuple(saturated.ln_phi),
        p_sat=tuple(from_si(value, "Pa") for value in saturated.p_sat),
        warnings=tuple(warnings),
    )

File: crates/azoth-eos/src/ge_uniquac_phase.rs
//! `eos.ge_uniquac_phase` - the fugacity coefficients of a UNIQUAC
//! activity-coefficient liquid.
//!
//! Spec: `specs/models/eos/ge_uniquac_phase.toml`. A *direct* model: no iteration, so no
//! `algorithm` block.
//!
//! NeqSim's `PhaseGEUniquac` is the port source. Its liquid is `ComponentGE.fugcoef`'s
//! `phi_i = gamma_i P0_i / P`, the same composition the other GE phases make - every GE
//! phase inherits that method and none overrides it - with `gamma_i` from UNIQUAC.
//! Both halves are already ported and this is their composition rather than a third
//! piece of physics.
//!
//! NeqSim's own UNIQUAC `gamma` does not exist: `ComponentGEUniquac.getGamma` returns
//! `0.0` and its formula is commented out, so unlike the other phases this one has no
//! oracle for its composition. The `r`/`q` half is checked instead, against
//! `ComponentGEUnifac.getR`/`getQ`.

use azoth_core::units::pascals;
use azoth_core::{AzothError, Result, apply_checks};

use crate::databank::{GeUniquacPhaseParameters, UniquacParameters};
use crate::ge_phase::ge_fugacities;
use crate::model_gen;
use crate::results::GeUniquacPhaseResult;

/// The fugacity coefficients of a liquid whose non-ideality is UNIQUAC's.
///
/// `phi_i = gamma_i P0_i / P`, which is NeqSim's `ComponentGE.fugcoef` for this phase.
/// `gamma_i` comes from UNIQUAC on `params`' `r` and `q` with the caller's `aij`;
/// `P0_i` is Antoine's correlation on the component's own coefficients at `T`. `params`
/// is resolved by name through [`crate::databank::ge_uniquac_phase_parameters`]; `x` is
/// checked rather than renormalised.
///
/// `aij` is `N x N`, in kelvin, directional with a zero diagonal - the same argument
/// `eos.uniquac_activity_coefficients` takes, and for the same reason: no upstream table
/// carries it.
///
/// # Errors
/// * [`AzothError::InvalidInput`] if `params` and `x` disagree in length, if `x` is not a
///   composition, or if `aij` is not an `N x N` matrix with a zero diagonal.
/// * [`AzothError::OutOfRange`] if `T` or `P` is not positive.
#[allow(non_snake_case)] // `T`, `P` and `x` are the symbols in the chemistry
pub fn ge_uniquac_phase(
    params: &GeUniquacPhaseParameters,
    T: f64,
    P: f64,
    x: &[f64],
    aij: &[Vec<f64>],
) -> Result<GeUniquacPhaseResult> {
    let spec = &model_gen::GE_UNIQUAC_PHASE_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "T" => Some(T),
            "P" => Some(P),
            _ => None,
        },
        &mut warnings,
    )?;

    let n = x.len();
    if n == 0 {
        return Err(AzothError::invalid_input(
            "x",
            "a mixture of zero components has no fugacity coefficient",
        ));
    }
    if params.antoine.len() != n {
        return Err(AzothError::invalid_input(
            "components",
            format!(
                "the phase has {} component(s) but `x` has {n} entries",
                params.antoine.len()
            ),
        ));
    }
    if let Some(bad) = x.iter().position(|&value| value < 0.0) {
        return Err(AzothError::invalid_input(
            "x",
            format!(
                "x[{bad}] is {} but a mole fraction cannot be negative",
                x[bad]
            ),
        ));
    }
    let sum: f64 = x.iter().sum();
    if (sum - 1.0).abs() > 1.0e-09 {
        return Err(AzothError::invalid_input(
            "x",
            format!(
                "the mole fractions sum to {sum}, not to one. Renormalising them here \
                 would make a composition error invisible in every number downstream, \
                 so it is refused instead"
            ),
        ));
    }
    if aij.len() != n || aij.iter().any(|row| row.len() != n) {
        return Err(AzothError::invalid_input(
            "aij",
            format!(
                "the interaction matrix is {} x {} but the mixture has {n} components",
                aij.len(),
                aij.first().map_or(0, Vec::len)
            ),
        ));
    }

    // The activity coefficients, from the same r and q and the same interaction matrix
    // `eos.uniquac_activity_coefficients` uses - resolved through the same function, so the
    // two cannot disagree.
    let uniquac = UniquacParameters {
        r: params.r.clone(),
        q: params.q.clone(),
    };
    let activity =
        crate::uniquac_activity_coefficients::uniquac_activity_coefficients(&uniquac, T, x, aij)?;
    warnings.extend(activity.warnings);
    let gamma = activity.gamma;

    // The saturation pressures and the composition, in the one place every GE phase
    // shares them.
    let saturated = ge_fugacities(&gamma, &params.antoine, T, P)?;
    warnings.extend(saturated.warnings);

    let ln_gamma: Vec<f64> = gamma.iter().map(|g| g.ln()).collect();
    Ok(GeUniquacPhaseResult {
        gamma,
        ln_gamma,
        ln_phi: saturated.ln_phi,
        p_sat: saturated.p_sat.into_iter().map(pascals).collect(),
        warnings,
    })
}
